### PR TITLE
add syntax highlighting for output

### LINKED ISSUE #39: syntax highlight
add option to syntax highlight the output as discussed here: https://twitter.com/Aguay_val/status/1287298579355381760

### PR DESCRIPTION
**Pull request by karuppiah7890**
fixes #39

- [x] Add code for syntax highlight for top level / root command
- [x] Add test for syntax highlight for top level / root command
- [x] Add code for syntax highlight for get sub command
- [x] Add test for syntax highlight for get sub command

### CODE REVIEW

**Comment by karuppiah7890** [comment]
Consider this a draft. Of course we can first discuss and make any necessary changes and then merge :)

**Comment by karuppiah7890** [comment]
It currently looks like this

Screen Shot 2020-12-30 at 7 51 05 PM

**Comment by karuppiah7890** [comment]
With yaml, the output looks nice, but with JSON, it doesn't look that great actually, even though JSON is only a subset of YAML

**Comment by karuppiah7890** [comment]
And I haven't added any unit or E2E test. I was wondering if we should test it and what to test and how to do it.

Should we add just E2E test? Since it's more CLI output related. I can't think of anything related to unit test, but I can see that there are some unit tests around output. We could do that.

Any thoughts on testing?

**Comment by AssafKatz3** [comment]
> 
> 
> And I haven't added any unit or E2E test. I was wondering if we should test it and what to test and how to do it.
> 
> Should we add just E2E test? Since it's more CLI output related. I can't think of anything related to unit test, but I can see that there are some unit tests around output. We could do that.
> 
> Any thoughts on testing?

1. Check that the option has impact (different output) for JSON & YAML with/without this option
2. Check that error code is consistent with and without this option
3. Check that writing it twice pass correctly
4. Check that the first line of both YAML & JSON output contains color sequence
5. Check the color of YAML by reread the output and checking that all color sequences are in place (It is problematic for JSON since it will fail for numbers)

**Comment by karuppiah7890** [comment]
Could you explain more about what points 2 and 3 mean? I didn't understand them properly

**Comment by AssafKatz3** [comment]
> 
> 
> Could you explain more about what points 2, 3 and 5 mean? I didn't understand them properly

2 - Ensure that if the command failed without color, it is failing with color also
3 - Call command with "-c -c" and see that it is working
5 - Call command with color, read the output and compare it to what should be the output (with color character sequences)

**Comment by karuppiah7890** [comment]
What's the benefit of 4 (single line check) if 5 does a complete check?

**Comment by AssafKatz3** [comment]
> 
> 
> What's the benefit of 4 (single line check) if 5 does a complete check?

4 is working for JSON also, I agree that for YAML is redundant if you will do also 5.

**Comment by karuppiah7890** [comment]
About 2 - Why would the error code change because of this option? Only when highlighting has some issues (error), there will be a non-zero exit code

**Comment by AssafKatz3** [comment]
> 
> 
> About 2 - Why would the error code change because of this option? Only when highlighting has some issues (error), there will be a non-zero exit code

It shouldn't - this is reason why need such test (I had such bug 20 years ago and it was really nagging)

**Comment by karuppiah7890** [comment]
Makes sense. I'll add it.

But I was just wondering, so, the test strategy is -

For every feature flag that's present, test if the usage of any flag causes the exit code to change in an unexpected manner? Doesn't that sound tedious? Or it could be done as some sort of parameterized test, but still that would be a lot of combinations depending on the number of flags and possible error situations

**Comment by AssafKatz3** [comment]
> This project doesn't have much flags, so all good. Just wondering in general 😅

I think that this is question for @itaysk and should be documented somewhere.

**Comment by karuppiah7890** [comment]
Cool. I thought you would have an answer since you suggested the strategy

**Comment by karuppiah7890** [comment]
Another thing is, I have only written code for the root command. I'm yet to understand and try out the "get" sub command. If it needs the color, we need to add for it too

**Comment by karuppiah7890** [comment]
Do we surely need the full output check? I think we can trust the library and just check that the option works - that is, invokes the library, by just checking the first line

**Comment by AssafKatz3** [comment]
> 
> 
> Do we surely need the full output check? I think we can trust the library and just check that the option works - that is, invokes the library, by just checking the first line

I am not sure this why I split it to two different tests.

**Comment by karuppiah7890** [comment]
I have finished all the stuff for this PR. cc @itaysk

**Comment by karuppiah7890** [comment]
You can review and let me know how it looks and if there are any more changes to be done :)

**Comment by AssafKatz3** [comment]
@karuppiah7890 
[The library that this code rely on](https://github.com/andreazorzetto/yh/blob/master/highlight/highlight.go), which in turn depended on [another library which isn't working well on Windows](https://github.com/logrusorgru/aurora#windows). I didn't check it personally, but it is something to check (manually at least) :-(

**Comment by karuppiah7890** [comment]
@AssafKatz3 cool, I'll check it out! Currently we still don't have support for windows however (no artifacts) but yeah, I'll be sure to check it out

File: cmd/cmd.go [hunk 1]
```diff
@@ -122,11 +128,29 @@ kubectl neat get -- svc -n default myservice --output json`,
...
-		cmd.Println(string(out))
+		err = printOutput(cmd, out)
+		if err != nil {
+			return err
+		}
 		return nil
 	},
 }
 
+func printOutput(cmd *cobra.Command, out []byte) error {
```
Comment (itaysk) [inline cmd/cmd.go:139]: instead of passing the entire cmd, can you please in just `cmd.OutOrStdout()` as `io.writer`?

File: cmd/cmd.go [hunk 2]
```diff
@@ -122,11 +128,29 @@ kubectl neat get -- svc -n default myservice --output json`,
...
+	if *color {
+		outreader := bytes.NewReader(out)
+		highlightedOutput, err := highlight.Highlight(outreader)
+		if err != nil {
+			return err
+		}
+		cmd.Println(highlightedOutput)
```
Comment (itaysk) [inline cmd/cmd.go:146]: please rebase and use `Print` based on https://github.com/itaysk/kubectl-neat/pull/57

**Review by itaysk — CHANGES_REQUESTED** [review]

**Comment by karuppiah7890** [comment]
Apparently the library may not work on Windows - https://github.com/itaysk/kubectl-neat/pull/52#issuecomment-752850500 . I'm not working on this PR now, so I'll close it. I hope new contributors will have an opportunity to work on it and also check support for windows based on #6 , #53

**Comment by karuppiah7890** [comment]
I'll resume working on this soon

**Pull request closed without merging**